File: scripts/xmlToJson.js
/* G. Hemingway Copyright @2014
 * Convert a CAD model (per the STEPTOOLS defined XML spec) into a JSON spec model
 */

var fs = require("fs"),
    _ = require("underscore"),
    cp = require("child_process"),
    xml2js = require("xml2js"),
    libxmljs = require("libxmljs");

var readTime = 0, parseTime = 0, translateTime = 0, writeTime = 0;
var config = {
    indexPoints: true,
    indexNormals: true,
    indexColors: true,
    compressColors: true,
    roundPrecision: 2
};

/***********************************************************************/


var roundFloat = function(val, precision) {
    if (!precision) return val;
    var factor = Math.pow(10, precision);
    return Math.round(val * factor);// / factor;
};

/***********************************************************************/


var translateIndex = function(doc) {
    // Return the full JSON
    translateTime = Date.now();
    var data = {
        root:       doc.root().attr("root").value(),
        products:   _.map(doc.find("//product"), translateProduct),
        shapes:     _.map(doc.find("//shape"), translateShape),
        shells:     _.map(doc.find("//shell"), translateShell),
        annotations:_.map(doc.find("//annotation"), translateAnnotation)
    };
    translateTime = Date.now() - translateTime;
    return data;
};

var showIndex = function(data) {
    var externalShells = _.pluck(data.shells, "href");
    var externalAnnotations = _.pluck(data.annotations, "href");
    console.log("\tProducts: " + data.products.length);
    console.log("\tShapes: " + data.shapes.length);
    console.log("\tAnnotations: " + data.annotations.length);
    console.log("\tExternal Annotations: " + externalAnnotations.length);
    console.log("\tShells: " + data.shells.length);
    console.log("\tExternal Shells: " + externalShells.length);
};

var translateProduct = function(product) {
    var data = {
        "id": product.attr("id").value(),
        "step": product.attr("step").value(),
        "name": product.attr("name").value()
    };
    // Add children, if there are any
    if (product.attr("children")) {
        data.children = product.attr("children").value().split(" ");
    }
    // Add shapes, if there are any
    if (product.attr("shape").value()) {
        data.shapes = product.attr("shape").value().split(" ");
    }
    return data;
};

var setTransform = function(transform) {
    // Look for identity transforms
    if (transform === "1 0 0 0 0 1 0 0 0 0 1 0 0 0 0 1") return "I";
    // Otherwise, turn this into an array of float values
    return transform.split(" ").map(function(val) {
        return parseFloat(val);
    })
};

var translateShape = function(shape) {
    // Base Shape JSON
    var data = {
        "id": shape.attr("id").value(),
        "shells": [],
        "annotations": [],
        "children": []
    };
    // Add children, if there are any
    _.forEach(shape.find("child"), function(child) {
        data.children.push({
            "ref": child.attr("ref").value(),
            "xform": setTransform(child.attr("xform").value())
        });
    });
    // Add child annotations
    if (shape.attr("annotation")) {
        data.annotations = shape.attr("annotation").value().split(" ");
    }
    // Terminal Shape JSON
    if (shape.attr("shell")) {
        data.shells = shape.attr("shell").value().split(" ");
    }
    return data;
};

var translateAnnotation = function(annotation) {
    var data = {
        "id": annotation.attr("id").value()
    };
    // Is this a non-terminal annotation
    if (annotation.attr("href")) {
        data.href = annotation.attr("href").value().replace("xml", "json");
    // Otherwise, add all those lines
    } else {
        translateTime = Date.now();
        data.lines = _.map(annotation.find("polyline"), function(polyline) {
            var points = [];
            _.forEach(polyline.find("p"), function(line) {
                _.forEach(line.attr("l").value().split(" "), function(val) {
                    points.push(parseFloat(val));
                });
            });
            return points;
        });
        translateTime = Date.now() - translateTime;
    }
    return data;
};

/***********************************************************************/

var indexShellPoints = function(data) {
    var numPoints = data.points.length;
    data.pointsIndex = [];
    for (var i = 0; i < numPoints; i++) {
        var val = roundFloat(data.points[i], config.roundPrecision);
        // See if this norm is already known
        var index = data.values.indexOf(val);
        if (index === -1) {
            index = data.values.push(val) - 1;
        }
        data.pointsIndex.push(index);
    }
    delete data.points;
};

var indexShellNormals = function(data) {
    var numNormals = data.normals.length;
    var indexArray = [];
    for (var i = 0; i < numNormals; i++) {
        var val = roundFloat(data.normals[i], config.roundPrecision);
        // See if this norm is already known
        var index = data.values.indexOf(val);
        if (index === -1) {
            index = data.values.push(val) - 1;
        }
        indexArray.push(index);
    }
    data.normalsIndex = indexArray;
    delete data.normals;
};

var indexShellColors = function(data) {
    var numColors = data.colors.length;
    var indexArray = [];
    for (var i = 0; i < numColors; i++) {
        var val = roundFloat(data.colors[i], config.roundPrecision);
        // See if this norm is already known
        var index = data.values.indexOf(val);
        if (index === -1) {
            index = data.values.push(val) - 1;
        }
        indexArray.push(index);
    }
    data.colorsIndex = indexArray;
    delete data.colors;
};

var compressShellColors = function(data) {
    var numTuples = data.colorsIndex.length / 3;
    data.colorsData = [];
    var start = 0;
    var last = [
        data.colorsIndex[0],
        data.colorsIndex[1],
        data.colorsIndex[2]
    ];
    // Short list comparison
    function arraysIdentical(a, b) {
       return (a[0] === b[0] && a[1] === b[1] && a[2] === b[2]);
    }
    // Compress the rest
    for (var tuple = 1; tuple < numTuples; tuple++) {
        var index = tuple * 3;
        var tmp = [
            data.colorsIndex[index],
            data.colorsIndex[index + 1],
            data.colorsIndex[index + 2]
        ];
        // Is this a new block
        if (!arraysIdentical(last, tmp)) {
            data.colorsData.push({
                data: last,
                duration: tuple - start
            });
            start = tuple;
            last = tmp;
        }
    }
    // Push the final color block
    data.colorsData.push({
        data: last,
        duration: numTuples - start
    });
    // Remove the colorsIndex
    delete data.colorsIndex;
//    console.log(JSON.stringify(data.colorsData));
};

/***********************************************************************/

var translateShell = function(shell) {
    // Do href here
    if (shell.attr("href")) {
        return {
            "id": shell.attr("id").value(),
            "size": parseInt(shell.attr("size").value()),
            "bbox": shell.attr("bbox").value().split(" ").map(function(val) { return parseFloat(val); }),
            "href":  shell.attr("href").value().replace("xml", "json")
        };
    // Convert XML point/vert/color to new way
    } else {
        translateTime = Date.now();
        var points = loadPoints(shell.find("verts"));
        var defaultColor = parseColor("7d7d7d");
        if (shell.attr("color")) {
            defaultColor = parseColor(shell.attr("color").value());
        }
        var data = {
            "id": shell.attr("id").value(),
            "size": 0,
            "points": [],
            "normals": [],
            "colors": []
        };
        _.forEach(shell.find("facets"), function(facet) {
            var color = _.clone(defaultColor);
            if (facet.attr("color")) {
                color = parseColor(facet.attr("color").value());
            }
            _.forEach(facet.find("f"), function(f) {
                // Get every vertex index and convert using points array
                var indexVals = f.attr("v").value().split(" ");
                var index0 = parseInt(indexVals[0]) * 3;
                var index1 = parseInt(indexVals[1]) * 3;
                var index2 = parseInt(indexVals[2]) * 3;

                data.points.push(parseFloat(points[index0]));
                data.points.push(parseFloat(points[index0 + 1]));
                data.points.push(parseFloat(points[index0 + 2]));
                data.points.push(parseFloat(points[index1]));
                data.points.push(parseFloat(points[index1 + 1]));
                data.points.push(parseFloat(points[index1 + 2]));
                data.points.push(parseFloat(points[index2]));
                data.points.push(parseFloat(points[index2 + 1]));
                data.points.push(parseFloat(points[index2 + 2]));

                // Get the vertex normals
                var norms = f.find("n");
                var normCoordinates = norms[0].attr("d").value().split(" ");
                data.normals.push(parseFloat(normCoordinates[0]));
                data.normals.push(parseFloat(normCoordinates[1]));
                data.normals.push(parseFloat(normCoordinates[2]));
                normCoordinates = norms[1].attr("d").value().split(" ");
                data.normals.push(parseFloat(normCoordinates[0]));
                data.normals.push(parseFloat(normCoordinates[1]));
                data.normals.push(parseFloat(normCoordinates[2]));
                normCoordinates = norms[2].attr("d").value().split(" ");
                data.normals.push(parseFloat(normCoordinates[0]));
                data.normals.push(parseFloat(normCoordinates[1]));
                data.normals.push(parseFloat(normCoordinates[2]));

                // Get the vertex colors
                data.colors.push(color.r);
                data.colors.push(color.g);
                data.colors.push(color.b);
                data.colors.push(color.r);
                data.colors.push(color.g);
                data.colors.push(color.b);
                data.colors.push(color.r);
                data.colors.push(color.g);
                data.colors.push(color.b);
            });
        });
        // Set the point data size
        data.size = data.points.length / 9;
        if (config.indexPoints || config.indexNormals || config.indexColors) {
            data.values = [];
            if (config.roundPrecision) {
                data.precision = config.roundPrecision;
            }
        }
        // Should we index the normals
        if (config.indexPoints) {
            indexShellPoints(data);
        }
        // Should we index the normals
        if (config.indexNormals) {
            indexShellNormals(data);
        }
        // Should we index the colors
        if (config.indexColors) {
            indexShellColors(data);
            // Should we hyper-compress the colors
            if (config.compressColors) {
                compressShellColors(data);
            }
        }
        translateTime = Date.now() - translateTime;
        return data;
    }
};

function parseColor(hex) {
    var cval = parseInt(hex, 16);
    return {
        r: ((cval >>16) & 0xff) / 255,
        g: ((cval >>8) & 0xff) / 255,
        b: ((cval >>0) & 0xff) / 255
    };
}

function loadPoints(verts) {
    // Load all of the point information
    var points = [];
    _.forEach(verts, function(vert) {
        _.forEach(vert.find("v"), function(v) {
            var coords = v.attr("p").value().split(" ");
            points.push(coords[0]);
            points.push(coords[1]);
            points.push(coords[2]);
        });
    });
    return points;
}

/*************************************************************************/

function XMLTranslator() {
    this.parser = new xml2js.Parser();
}

XMLTranslator.prototype.parse = function(dir, filename, callback) {
    var self = this;
    this.pathPrefix = dir + "/";
    var rootPath = this.pathPrefix + filename;
    // Read the root file
    readTime = Date.now();
    fs.readFile(rootPath, function(err, doc) {
        if (err) {
            callback(err);
        } else {
            parseTime = Date.now();
            readTime = parseTime - readTime;
            var results = libxmljs.parseXmlString(doc);
            parseTime = Date.now() - parseTime;
            callback(results.errors[0], results);
        }
    });
};

XMLTranslator.prototype.write = function(directory, filename, data, callback) {
    var path = directory + "/" + filename.replace("xml", "json");
    // Write the object to file
    writeTime = Date.now();
    fs.writeFile(path, JSON.stringify(data), function(err) {
        writeTime = Date.now() - writeTime;
        if (callback) callback(err, data);
    });
};

/*************************************************************************/

function showTimes() {
    console.log(argv.f + ": " +
        "(R: " + readTime + ", " +
        "P: " + parseTime + ", " +
        "T: " + translateTime + ", " +
        "W: " + writeTime + ")");
}

var argv = require('optimist')
    .default('d', '.')
    .argv;

var translator = new XMLTranslator();
// Translate the requested model file
translator.parse(argv.d, argv.f, function(err, data) {
    if (err) {
        console.log(err);
        return;
    }
    var root = data.root();
    // Is this is an assembly
    switch (root.name()) {
        case "step-assembly":
        case "assembly":
            translator.write(argv.d, argv.f, translateIndex(data), function(err, data) {
//                showIndex(data);
                // What is external
                var externalShells = _.pluck(data.shells, "href");
                var externalAnnotations = _.pluck(data.annotations, "href");
                // Push jobs to the workers
                _.forEach(externalShells, function(shell) {
                    shell = shell.replace("json", "xml");
                    cp.fork("scripts/xmlToJson.js", ["-d", argv.d, "-f", shell]);
                });
                _.forEach(externalAnnotations, function(annotation) {
                    annotation = annotation.replace("json", "xml");
                    cp.fork("scripts/xmlToJson.js", ["-d", argv.d, "-f", annotation]);
                });
                showTimes();
            });
            break;
        case "shell":
            translator.write(argv.d, argv.f, translateShell(data.root()), function(err, data) {
                showTimes();
            });
            break;
        case "annotation":
            translator.write(argv.d, argv.f, translateAnnotation(data.root()), function(err, data) {
                showTimes();
            });
            break;
        default:
            console.log("Unknown XML file type");
    }
});

File: scripts/xmlToJson.py
#!/usr/bin/env python

# L. Howard Copyright @2014
# Convert a CAD model (per the STEPtools defined XML spec)
# into a JSON spec model
# Derived from Javascript version at
# https://github.com/ghemingway/cad.js/blob/master/scripts/xmlToJson.js

import argparse
from datetime import datetime
import json
import math
from multiprocessing import cpu_count, Process, Queue
from operator import itemgetter
import os
from os.path import join
import re
import sys
import time
import xml.etree.cElementTree as ET

import logging
logging.basicConfig(
    format='%(asctime)s %(levelname)s:%(message)s', level=logging.DEBUG)
LOG = logging.getLogger(__name__)

# defaults and constants
DEFAULT_COLOR = "7d7d7d"
IDENTITY_TRANSFORM = "1 0 0 0 0 1 0 0 0 0 1 0 0 0 0 1"
SHELL_REGEX = re.compile("shell_(.*?).json")

#------------------------------------------------------------------------------

CONFIG = {
    'indexPoints': True,
    'indexNormals': True,
    'compressColors': True,
    'roundPrecision': 2
}


def round_float(val, precision):
    """floating point rounder"""
    if not precision:
        return val
    factor = math.pow(10, precision)
    return int(round(val * factor))


#------------------------------------------------------------------------------

def translate_index(doc):
    """Returns the full JSON"""
    return {
        'root': doc.attrib['root'],
        'products': [translate_product(x) for x in doc.iter('product')],
        'shapes': [translate_shape(x) for x in doc.iter('shape')],
        'shells': [translate_shell(x) for x in doc.iter('shell')],
        'annotations': [translate_annotation(x) for x in doc.iter('annotation')]
    }


def translate_product(product):
    """Translates a product"""
    data = {
        'id': product.attrib['id'],
        'step': product.attrib.get('step', ""),
        'name': product.attrib['name']
    }
    # Add children, if there are any
    if product.attrib.get('children'):
        data['children'] = product.attrib['children'].split(" ")
    # Add shapes, if there are any
    if product.attrib.get('shape'):
        data['shapes'] = product.attrib['shape'].split(" ")
    return data


def set_transform(transform):
    """Sets a transform"""
    return ("I" if transform == IDENTITY_TRANSFORM else
            [float(x) for x in transform.split(" ")])


def translate_shape(shape):
    """Translates a shape"""
    data = {
        'id': shape.attrib['id'],
        # "unit": shape.attrib['unit'],
    }
    data.update({x: [] for x in ('shells', 'annotations', 'children')})
    for child in shape.iter('child'):
        # Add children, if there are any
        data['children'].append({
            'ref': child.attrib['ref'],
            'xform': set_transform(child.attrib['xform'])
        })
    # Add child annotations
    if shape.attrib.get('annotation'):
        data['annotations'] = shape.attrib['annotation'].split(" ")
    # Terminal Shape JSON
    if shape.attrib.get('shell'):
        data['shells'] = shape.attrib['shell'].split(" ")
    return data


def translate_annotation(annotation):
    """Translates an annotation"""
    data = dict(id=annotation.attrib['id'])
    if 'href' in annotation.attrib:
        data['href'] = annotation.attrib['href'].replace("xml", "json")
    else:
        data['lines'] = []
        for polyline in annotation.iter('polyline'):
            points = []
            for p in polyline.iter('p'):
                points.extend([float(x) for x in p.attrib['l'].split(" ")])
            data['lines'].append(points)
    return data


#------------------------------------------------------------------------------

def make_index(data, ikey, ranger=None):
    """Create indexes, an abstraction of indexing functions in the original"""
    if ranger is None:
        ranger = xrange(len(data[ikey]))
    indexes = data[ikey + "Index"] = []
    values = data['values']
    for i in ranger:
        val = round_float(data[ikey][i], CONFIG['roundPrecision'])
        if val not in values:
            values[val] = len(values)
        indexes.append(values[val])
    del data[ikey]

# The following functions are named indexShellxxx in the original
# The renaming aligns them with settings in the indexing CONFIG
indexPoints = lambda d: make_index(d, 'points')
indexNormals = lambda d: make_index(d, 'normals')


def compress_shell_colors(data):
    """Color compression"""
    num_tuples = len(data['colors']) / 3
    data['colorsData'] = []
    start = 0
    last = [data['colors'][x] for x in xrange(3)]
    # Short list comparison
    arrays_identical = lambda a, b: all([a[x] == b[x] for x in xrange(3)])
    # Compress the rest
    for tupl in xrange(num_tuples):
        index = tupl * 3
        tmp = [data['colors'][index + x] for x in xrange(3)]
        # Is this a new block?
        if not arrays_identical(last, tmp):
            data['colorsData'].append(dict(data=last, duration=tupl - start))
            start = tupl
            last = tmp
    # append the final color block
    data['colorsData'].append(dict(data=last, duration=num_tuples - start))
    # remove the colors index
    del data['colors']


#------------------------------------------------------------------------------

def translate_shell(shell):
    """Translates a shell"""
    if 'href' in shell.attrib:
        # Do href here
        return {
            'id': shell.attrib['id'],
            'size': int(shell.attrib['size']),
            'bbox': [float(x) for x in shell.attrib['bbox'].split(" ")],
            'href': shell.attrib['href'].replace("xml", "json")
        }
    else:
        # Convert XML point/vert/color to new way
        points = load_points(shell.iter("verts"))
        default_color = parse_color(shell.attrib.get('color', DEFAULT_COLOR))
        data = dict(id=shell.attrib['id'], size=0)
        data.update({x: [] for x in ('points', 'normals', 'colors')})
        for facet in shell.iter('facets'):
            color = default_color
            if 'color' in facet.attrib:
                color = parse_color(facet.attrib['color'])
            for f in facet.iter('f'):
                # Get every vertex index and convert using points array
                index_vals = f.attrib['v'].split(" ")
                for i in range(3):
                    ival = int(index_vals[i]) * 3
                    data['points'].append(float(points[ival]))
                    data['points'].append(float(points[ival + 1]))
                    data['points'].append(float(points[ival + 2]))

                # Get the vertex normals
                norms = [x for x in f.iter('n')]
                for i in range(3):
                    norm_coordinates = norms[i].attrib['d'].split(" ")
                    for j in range(3):
                        data['normals'].append(float(norm_coordinates[j]))

                # Get the vertex colors
                for i in range(3):
                    for c in ('r', 'g', 'b'):
                        data['colors'].append(color[c])

        data['size'] = len(data['points']) / 9
        indexing = [x for x in CONFIG if x.startswith('index') and CONFIG[x]]
        for i in indexing:
            data['precision'] = CONFIG['roundPrecision']
            if 'values' not in data:
                data['values'] = {}
            globals()[i](data)
        if indexing:
            sorted_vals = sorted(data['values'].items(), key=itemgetter(1))
            data['values'] = map(itemgetter(0), sorted_vals)
        if CONFIG.get('compressColors'):
            compress_shell_colors(data)
        return data


def parse_color(hex_color):
    """Parse color values"""
    cval = int(hex_color, 16)
    x = lambda b: ((cval >> b) & 0xff) / 255.0
    return {k: x(v) for k, v in dict(r=16, g=8, b=0).iteritems()}


def load_points(verts):
    """Load all of the point information"""
    points = []
    for vert in verts:
        for v in vert:
            points.extend(v.attrib['p'].split(" "))
    return points


#------------------------------------------------------------------------------

class WorkerBase(Process):
    """Base class for Workers"""

    def __init__(self, queue, exceptions):
        Process.__init__(self)
        self.queue = queue
        self.exceptions = exceptions

    def report_exception(self, job, reason):
        """Report a job exception"""
        info = dict(reason=reason)
        info.update(job)
        self.exceptions.put(info)

    def run(self):
        raise NotImplementedError


class BatchWorker(WorkerBase):
    """Worker process for parallelized shell batching"""

    def run(self):
        """Process jobs"""
        while True:
            job = self.queue.get()
            if job is None:
                break
            # process shells
            batch = {'shells': []}
            indexed = CONFIG['indexPoints'] or CONFIG['indexNormals']
            reindex = job['reindex'] and indexed
            if reindex:
                batch['values'] = {}
            for s in job['shells']:
                try:
                    with open(join(job['path'], s)) as f:
                        shell = json.load(f)
                    sid = SHELL_REGEX.match(s).group(1)
                    shell['id'] = sid
                    if reindex:
                        imap = {}
                        for i, value in enumerate(shell['values']):
                            if value not in batch['values']:
                                batch['values'][value] = len(batch['values'])
                            imap[i] = batch['values'][value]
                        del shell['values']
                        for item in ('points', 'normals'):
                            idx = item + 'Index'
                            if idx in shell:
                                shell[idx] = [imap[x] for x in shell[idx]]
                    batch['shells'].insert(0, shell)
                except Exception as e:
                    reason = "Error batching shell '{}': {}".format(s, e)
                    self.report_exception(job, reason)
                    continue
            # transform values to list
            if reindex:
                sorted_v = sorted(batch['values'].items(), key=itemgetter(1))
                batch['values'] = map(itemgetter(0), sorted_v)
            # write batch
            out_path = join(job['path'], job['name'] + ".json")
            try:
                with open(out_path, "w") as f:
                    json.dump(batch, f)
            except Exception as e:
                reason = "Unable to output JSON '{}': {}.".format(out_path, e)
                self.report_exception(job, reason)


class TranslationWorker(WorkerBase):
    """Worker process for parallelized translation"""

    def run(self):
        """Process jobs"""
        while True:
            job = self.queue.get()
            if job is None:
                break
            try:
                path = job['path']
                tree = ET.parse(path)
                root = tree.getroot()
            except Exception as e:
                fpath = job.get('path', '?')
                reason = "Unable to parse XML file '{}': {}.".format(fpath, e)
                self.report_exception(job, reason)
                continue
            try:
                data = job['translator'](root)
            except Exception as e:
                reason = "Translation failure: '{}'.".format(e)
                self.report_exception(job, reason)
                continue
            out_path = os.path.splitext(path)[0] + ".json"
            try:
                with open(out_path, "w") as f:
                    json.dump(data, f)
            except Exception as e:
                reason = "Unable to output JSON '{}': {}.".format(out_path, e)
                self.report_exception(job, reason)


class XMLTranslator(object):
    """Translates STEP XML files to JSON"""

    def __init__(self, batches=None, reindex=None):
        self.batches = batches
        self.reindex = reindex

    @staticmethod
    def assign(batches, shell):
        """simple bin packing"""
        name, size = shell
        blist = batches.values()
        best = min([x['total_size'] for x in blist])
        selected = [x for x in blist if x['total_size'] == best][0]
        selected['total_size'] += size
        selected['shells'].append(name)

    def get_batches(self, shells):
        """assign shells to batches, leveling by size"""
        batches = {'batch%s' % i:  {'total_size': 0, 'shells': []}
                   for i in xrange(self.batches)}
        for shell in shells:
            self.assign(batches, shell)
        return batches

    def batch_shells(self, xml_dir):
        """Generates batched shell files"""
        is_shell = lambda x: SHELL_REGEX.match(x)
        size_of = lambda x: os.path.getsize(join(xml_dir, x))
        shells = [(x, size_of(x)) for x in os.listdir(xml_dir) if is_shell(x)]
        shells.sort(key=itemgetter(1), reverse=True)
        batches = self.get_batches(shells)

        # start workers and queue jobs
        queue = Queue()
        exceptions = Queue()
        count = min(cpu_count(), self.batches)
        workers = [BatchWorker(queue, exceptions) for _ in xrange(count)]
        for w in workers:
            w.start()

        # enqueue jobs
        for batch, info in batches.items():
            job = {'path': xml_dir, 'name': batch, 'shells': info['shells'],
                   'reindex': self.reindex}
            queue.put(job)

        # add worker termination cues
        for _ in workers:
            queue.put(None)

        # wait for completion
        while any([x.is_alive() for x in workers]):
            time.sleep(1)

        # report errors, if any
        has_errors = not exceptions.empty()
        while not exceptions.empty():
            info = exceptions.get()
            msg = "Error processing '{}': {}"
            LOG.error(msg.format(info['path'], info['reason']))

        if not has_errors:
            # report statistics
            msg = "Batch sizes.  Smallest: {} Largest: {} Avg: {}"
            sz = [x['total_size'] for x in batches.values()]
            LOG.debug(msg.format(min(sz), max(sz), sum(sz) / len(batches)))
            c = [len(x['shells']) for x in batches.values()]
            msg = "Batch shells.  Smallest: {} Largest: {} Avg: {}"
            LOG.debug(msg.format(min(c), max(c), sum(c) / len(batches)))
            shells_size = sum([size for name, size in shells])
            msg = "Shells.  Count: {} Total Size: {} bytes."
            LOG.debug(msg.format(len(shells), shells_size))
            regex = re.compile("batch[0-9]*.json")
            is_batch = lambda x: regex.match(x)
            sizes = [size_of(x) for x in os.listdir(xml_dir) if is_batch(x)]
            batches_size = sum(sizes)
            msg = "Batches.  Count: {} Total Size: {} bytes."
            LOG.debug(msg.format(len(sizes), batches_size))
            compression = float(batches_size) / float(shells_size)
            LOG.debug("Compression: {}".format(compression))

        return has_errors

    def translate(self, xml_dir, xml_index):
        """Process index XML and enqueue jobs for workers"""
        if not os.path.isdir(xml_dir):
            LOG.error("'{}' is not a directory.".format(xml_dir))
            return True
        index_path = join(xml_dir, xml_index)
        if not os.path.isfile(index_path):
            LOG.error("Unable to locate index file '{}'.".format(index_path))
            return True
        try:
            tree = ET.parse(index_path)
            root = tree.getroot()
        except Exception as e:
            LOG.exception("Unable to parse '{}'.".format(index_path))
            return True
        try:
            data = translate_index(root)
        except Exception as e:
            LOG.exception("Unable to translate index file.")
            return True

        pluck = lambda e, a: [x for x in data.get(e, []) if a in x]
        external_shells = pluck('shells', 'href')
        external_annotations = pluck('annotations', 'href')
        index_out = join(xml_dir, os.path.splitext(xml_index)[0] + ".json")

        LOG.debug("Writing new index file: " + index_out)
        LOG.debug("\tProducts: %s" % len(data.get('projects', [])))
        LOG.debug("\tShapes: %s" % len(data.get('shapes', [])))
        LOG.debug("\tAnnotations: %s" % len(data.get('annotations', [])))
        LOG.debug("\tExternal Annotations: %s" % len(external_annotations))
        LOG.debug("\tShells: %s" % len(data.get('shells', [])))
        num_shells = len(external_shells)
        LOG.debug("\tExternal Shells: %s" % num_shells)
        if self.batches and num_shells:
            if num_shells < self.batches:
                self.batches = 1
            LOG.debug("\tBatches: %s" % self.batches)
            data['batches'] = self.batches
        else:
            self.batches = 0

        try:
            with open(index_out, "w") as f:
                json.dump(data, f)
        except Exception as e:
            LOG.exception("Unable to write JSON file '{}'.".format(index_out))
            return True

        # start workers and queue jobs
        queue = Queue()
        exceptions = Queue()
        count = cpu_count()
        workers = [TranslationWorker(queue, exceptions) for _ in xrange(count)]
        for w in workers:
            w.start()

        xml_path = lambda p: join(xml_dir, os.path.splitext(p)[0] + ".xml")
        for annotation in external_annotations:
            queue.put({
                'type': "annotation",
                'path': xml_path(annotation['href']),
                'translator': translate_annotation
            })

        for shell in external_shells:
            queue.put({
                'type': "shell",
                'path': xml_path(shell['href']),
                'translator': translate_shell
            })

        # add worker termination cues
        for _ in workers:
            queue.put(None)

        # wait for completion
        while any([x.is_alive() for x in workers]):
            time.sleep(1)

        # report errors, if any
        has_errors = not exceptions.empty()
        while not exceptions.empty():
            info = exceptions.get()
            msg = "Error processing '{}': {}"
            LOG.error(msg.format(info['path'], info['reason']))

        if has_errors or not self.batches:
            return has_errors

        return self.batch_shells(xml_dir)

#------------------------------------------------------------------------------

if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        prog='xmlToJson.py',
        description="Translates STEP XML to JSON")
    parser.add_argument("dir", help="directory containing STEP XML")
    parser.add_argument("index", help="index file")
    h = "create batches of shells"
    parser.add_argument("-b", "--batches", type=int, default=0, help=h)
    h = "re-index when batching shells"
    parser.add_argument("-r", "--reindex", action="store_true", help=h)
    args = parser.parse_args()

    start_time = datetime.now()
    translator = XMLTranslator(args.batches, args.reindex)
    errors_in_translation = translator.translate(args.dir, args.index)
    dt = datetime.now() - start_time
    LOG.info("xmlToJson Elapsed time: {} secs".format(dt.seconds))
    sys.exit(1 if errors_in_translation else 0)
